### PR TITLE
fix: Complex secret expression loading

### LINKED ISSUE #1409: When using secrets in complex expressions directly in workflow actions, you may encounter error.
```yaml
# This FAILS in a workflow action
- ref: call_api
  action: core.http_request
  args:
    headers:
      Authorization: Basic ${{ FN.to_base64(SECRETS.zendesk.ZENDESK_EMAIL + "/token:" + SECRETS.zendesk.ZENDESK_API_TOKEN) }}
```

Error: `TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'`

However, the same expression works in two scenarios:
1. Inside a template action
2. When accompanied by a simple secret expression

##Templates Work

### Template Secret Declaration

Templates declare their required secrets upfront in the action definition:

```yaml
# templates/tools/zendesk/get_ticket.yml
type: action
definition:
  title: Get ticket
  namespace: tools.zendesk
  name: get_ticket
  secrets:
    - name: zendesk
      keys: ["ZENDESK_EMAIL", "ZENDESK_API_TOKEN"]  # ← Explicit declaration
  steps:
    - ref: encode_auth
      action: core.transform.reshape
      args:
        value: ${{ FN.to_base64(SECRETS.zendesk.ZENDESK_EMAIL + "/token:" + SECRETS.zendesk.ZENDESK_API_TOKEN) }}
```

### Template Execution Flow

1. **Secret Loading Phase** (BEFORE expression evaluation):
   ```python
   # In run_action_from_input() - executor/service.py
   async with RegistryActionsService.with_session() as service:
       reg_action = await service.get_action("tools.zendesk.get_ticket")
       action_secrets = await service.fetch_all_action_secrets(reg_action)
       # Returns: ["zendesk"] from the template's secret declaration
   
   secrets = await get_action_secrets(args=task.args, action_secrets=action_secrets)
   # Loads ALL keys for the "zendesk" secret
   ```

2. **Template Step Execution** (WITH secrets already loaded):
   ```python
   # In run_template_action() - executor/service.py
   context = {
       ExprContext.SECRETS: secrets_context,  # ← Already has zendesk.ZENDESK_EMAIL, etc.
   }
   
   for step in defn.steps:
       # Evaluate expressions with secrets available
       evaled_args = eval_templated_object(step.args, operand=context)
   ```

**Key Point**: Templates load secrets based on their **declaration**, not by scanning expressions.

### Workflow Action Secret Discovery

Direct workflow actions must discover their secrets by scanning the arguments:

```yaml
# In a workflow
- ref: my_action
  action: core.http_request
  args:
    headers:
      Authorization: Basic ${{ FN.to_base64(SECRETS.zendesk.ZENDESK_EMAIL + ...) }}
```

### The Discovery Problem

1. **Secret Discovery Phase** (regex-based):
   ```python
   # In get_action_secrets() - executor/service.py
   args_secrets = extract_templated_secrets(args)
   
   # Uses this regex pattern:
   SECRET_SCAN_TEMPLATE = re.compile(r"\${{\s*SECRETS\.(?P<secret>.+?)\s*}}")
   ```

2. **Pattern Matching Failure**:
   - Pattern expects: `${{ SECRETS.xxx }}`
   - But finds: `${{ FN.to_base64(SECRETS.xxx) }}`
   - The regex requires `SECRETS.` immediately after `${{`
   - Result: No secrets discovered → No secrets loaded

## The Workaround: Adding a Simple Expression Works

When you add a simple secret expression alongside the complex one:

```yaml
headers:
  Authorization: Basic ${{ FN.to_base64(SECRETS.zendesk.ZENDESK_EMAIL + ...) }}
  X-Init: ${{ SECRETS.zendesk.ZENDESK_EMAIL }}  # ← Simple expression
```

## Summary Table

| Scenario | Secret Discovery | Secret Loading | Expression Evaluation | Result |
|----------|-----------------|----------------|----------------------|---------|
| Template with complex expression | Via declaration | All declared secrets | Has secrets in context | ✅ Works |
| Direct action with complex expression only | Regex fails | No secrets loaded | Missing secrets | ❌ Fails |
| Direct action with complex + simple expression | Regex finds simple | Loads entire secret | Has secrets in context | ✅ Works |

### PR DESCRIPTION
**Pull request by topher-lo**
Fixes https://github.com/TracecatHQ/tracecat/issues/1409

## Checklist

- [x] Read [CONTRIBUTING.md](https://github.com/TracecatHQ/tracecat/blob/main/CONTRIBUTING.md).
- [x] PR title is short and non-generic (see previously [merged PRs](https://github.com/TracecatHQ/tracecat/pulls?q=is%3Apr+is%3Aclosed) for examples).
- [x] PR only implements a single feature or fixes a single bug.
- [ ] Tests passing (`uv run pytest tests`)?
- [ ] [Lint](https://docs.astral.sh/ruff/) / [pre-commits](https://pre-commit.com/) passing (`pre-commit run --all-files`)?

## Description

This PR fixes a bug where secrets used within complex expressions (e.g., `FN.to_base64(SECRETS.x + SECRETS.y)`) in direct workflow actions were not correctly discovered and loaded. This resulted in a `TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'` during expression evaluation.

The root cause was that the `extract_templated_secrets` function, responsible for discovering secrets, relied on a regex that only matched simple `SECRETS.<name>.<key>` expressions when they were the *entire* content of a `${{ ... }}` block. This prevented secrets from being loaded when they were nested inside function calls or concatenations.

The fix involves:
1.  **Refactoring `extract_templated_secrets`**: The function now first identifies all `${{ ... }}` template blocks. Then, for each identified expression, it uses a more granular regex to find all `SECRETS.<name>.<key>` occurrences *within* that expression. This ensures all secrets are detected regardless of their complexity or nesting.
2.  **Updating `SECRET_SCAN_TEMPLATE`**: The general `SECRET_SCAN_TEMPLATE` regex in `patterns.py` has also been updated to allow matching `SECRETS.<name>.<key>` anywhere inside a `${{ ... }}` block, improving its robustness for other potential uses.

Comprehensive unit tests have been added to cover the original bug scenario and various corner cases, including multiple secrets, whitespace, nested objects, and identifier variations.

## Related Issues

Problem description from conversation: `The problem is expression like this won't work. ${{ FN.to_base64(SECRETS.zendesk.ZENDESK_EMAIL + "/token:" + SECRETS.zendesk.ZENDESK_API_TOKEN) }}. You will get error Error: TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'`

## Screenshots / Recordings

N/A - Backend logic change.

## Steps to QA

1.  Create a workflow with a direct action (e.g., `core.http_request`).
2.  In the `args` of this action, use a complex expression involving secrets, such as:
    ```yaml
    headers:
      Authorization: Basic ${{ FN.to_base64(SECRETS.zendesk.ZENDESK_EMAIL + "/token:" + SECRETS.zendesk.ZENDESK_API_TOKEN) }}
    ```
3.  Ensure `SECRETS.zendesk.ZENDESK_EMAIL` and `SECRETS.zendesk.ZENDESK_API_TOKEN` are configured in your environment.
4.  Run the workflow.
5.  Verify that the workflow executes successfully without `TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'`.
6.  (Optional) Review the added unit tests in `tests/unit/test_expressions.py` and `tests/unit/test_executor_service.py` for comprehensive coverage.

---
<a href="https://cursor.com/background-agent?bcId=bc-006792b1-d697-43ea-88d5-2076a4bda881">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-cursor-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-cursor-light.svg">
    Open in Cursor
  </picture>
</a>
<a href="https://cursor.com/agents?id=bc-006792b1-d697-43ea-88d5-2076a4bda881">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-web-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-web-light.svg">
    Open in Web
  </picture>
</a>

    

---

## Summary by cubic
Fixes secret loading for complex template expressions so secrets nested in functions or string ops are discovered and loaded. Prevents NoneType errors in direct workflow actions.

- **Bug Fixes**
  - Updated secret extraction to scan inside ${{ ... }} and collect all SECRETS.<name>.<key> references, even when multiple or nested.
  - Relaxed regex patterns to match secrets anywhere within a template expression.
  - Added unit tests covering nested functions, multiple secrets, whitespace, dict keys/values, identifier variants, deduping, and non-template mentions.

### CODE REVIEW

**Comment by daryllimyt** [comment]
@cubic-dev-ai rereview

File: tracecat/expressions/eval.py [hunk 1]
```diff
@@ -75,22 +80,16 @@ def is_template_only(template: str) -> bool:
...
-
-    def operator(line: str) -> Any:
-        """Collect secrets from the templated string."""
-        for match in re.finditer(pattern, line):
-            secret = match.group("secret")
-            secrets.add(secret)
-
-    _eval_templated_obj_rec(templated_obj, operator)
-    return list(secrets)
+def extract_templated_secrets(templated_obj: Any) -> list[str]:
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/expressions/eval.py:83]: No — the comment isn’t valid for the current design.

Why it’s safe

Intentional behavior: extract_templated_secrets returns full secret paths (e.g., "name.KEY") by design via SecretPathExtractor which records the full “SECRETS.<name>.<key>” path (tracecat/expressions/eval.py:83, tracecat/expressions/core.py:208, tracecat/expressions/core.py:226).
Tests expect full paths: Unit tests assert the function returns full paths, including complex nested expressions (tests/unit/test_expressions.py:246, tests/unit/test_executor_service.py:282).
Downstream normalization: Consumers that fetch secrets accept either “name.KEY” or “name” and normalize to names during retrieval (splitting at “.”) so fetching is correct (tracecat/auth/sandbox.py:34, tracecat/auth/sandbox.py:135). get_action_secrets passes the set to AuthSandbox, which handles normalization (tracecat/executor/service.py:264).
About misclassification/incorrect fetching

OAuth classification in get_action_secrets checks endswith("_oauth") on extracted items, but that result isn’t used for fetching; secrets are fetched through AuthSandbox which handles both paths and names correctly (tracecat/executor/service.py:264).
If you specifically need names-only for some logic, use the registry-oriented extractor that returns base secret names (tracecat/expressions/core.py:201) or split the path at “.” to get the name.
Summary

Returning full paths is intentional, tested, and compatible with downstream fetching. There’s no breakage in current logic. If you need names-only behavior in a caller, split on “.” or use the name-only extractor.
